File: RadarSDK/RadarReplayBuffer.m
//
//  RadarReplayBuffer.m
//  RadarSDK
//
//  Copyright © 2023 Radar Labs, Inc. All rights reserved.
//

#import "Radar+Internal.h"
#import "RadarAPIClient.h"
#import "RadarReplayBuffer.h"
#import "RadarReplay.h"
#import "RadarLogger.h"
#import "RadarSettings.h"

static const int MAX_BUFFER_SIZE = 120; // one hour of updates
static const int MAX_BATCH_SIZE = 100;

@implementation RadarReplayBuffer {
    NSMutableArray<RadarReplay *> *mutableReplayBuffer;
    BOOL isFlushing;
    NSTimer *batchFlushTimer;
    NSDate *batchStartTime;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        mutableReplayBuffer = [NSMutableArray<RadarReplay *> new];
        isFlushing = NO;
        batchFlushTimer = nil;
        batchStartTime = nil;
    }
    return self;
}

+ (instancetype)sharedInstance {
    static dispatch_once_t once;
    static id sharedInstance;
    dispatch_once(&once, ^{
        sharedInstance = [self new];
    });
    return sharedInstance;
}

#pragma mark - Replay Buffer Methods

- (void)writeNewReplayToBuffer:(NSMutableDictionary *)replayParams {
    NSUInteger replayBufferLength = [mutableReplayBuffer count];
    if (replayBufferLength >= MAX_BUFFER_SIZE) {
        [self dropOldestReplay];
    }
    
    // add new replay to buffer
    RadarReplay *radarReplay = [[RadarReplay alloc] initWithParams:replayParams];
    [mutableReplayBuffer addObject:radarReplay];

    RadarSdkConfiguration *sdkConfiguration = [RadarSettings sdkConfiguration];
    if (sdkConfiguration.usePersistence) {
        NSData *replaysData;
        NSError *error;

        // if buffer length is above 50, remove every fifth replay from the persisted buffer
        if ([mutableReplayBuffer count] > 50) {
            NSMutableArray<RadarReplay *> *prunedBuffer = [NSMutableArray arrayWithCapacity:[mutableReplayBuffer count]];
            for (NSUInteger i = 0; i < mutableReplayBuffer.count; i++) {
                if ((i + 1) % 5 != 0) {
                    [prunedBuffer addObject:mutableReplayBuffer[i]];
                }
            }
            replaysData = [NSKeyedArchiver archivedDataWithRootObject:prunedBuffer requiringSecureCoding:YES error:&error];
        } else {
            replaysData = [NSKeyedArchiver archivedDataWithRootObject:mutableReplayBuffer requiringSecureCoding:YES error:&error];
        }
        if (error) {
            [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"Error archiving replays"]; 
            return;
        } else {
            [[NSUserDefaults standardUserDefaults] setObject:replaysData forKey:@"radar-replays"];
        }
    }
}

- (NSArray<RadarReplay *> *)flushableReplays {
    NSArray *flushableReplays = [mutableReplayBuffer copy];
    return flushableReplays;
}

- (void)flushReplaysWithCompletionHandler:(NSDictionary *_Nullable)replayParams
                        completionHandler:(RadarFlushReplaysCompletionHandler _Nullable)completionHandler {
    if (isFlushing) {
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"Already flushing replays"];
        if (completionHandler) {
            completionHandler(RadarStatusErrorServer, nil);
        }
        return;
    }

    isFlushing = YES;

    NSArray<RadarReplay *> *flushableReplays = [self flushableReplays];
    if ([flushableReplays count] == 0 && !replayParams) {
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"No replays to flush"];
        isFlushing = NO;
        if (completionHandler) {
            completionHandler(RadarStatusSuccess, nil);
        }
        return;
    }

    // get a copy of the replays so we can safely clear what was synced up
    NSMutableArray<RadarReplay *> *replaysArray = [NSMutableArray arrayWithArray:flushableReplays];
    
    NSMutableArray *replaysRequestArray = [RadarReplay arrayForReplays:replaysArray];

    // if we have a current track update, add it to the local replay list
    NSMutableDictionary *newReplayParams;
    if (replayParams) {
        newReplayParams = [replayParams mutableCopy];
        newReplayParams[@"replayed"] = @(YES);
        long nowMs = (long)([NSDate date].timeIntervalSince1970 * 1000);
        newReplayParams[@"updatedAtMs"] = @(nowMs);
        // remove the updatedAtMsDiff key because for replays we want to rely on the updatedAtMs key for the time instead
        [newReplayParams removeObjectForKey:@"updatedAtMsDiff"];
        [replaysRequestArray addObject:newReplayParams];
    }

    // log the replay count
    [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:[NSString stringWithFormat:@"Flushing %lu replays", (unsigned long)[replaysRequestArray count]]];

    [[RadarAPIClient sharedInstance] flushReplays:replaysRequestArray completionHandler:^(RadarStatus status, NSDictionary *_Nullable res) {
        if (status == RadarStatusSuccess) {
            // if the flush was successful, remove the replays from the buffer
            [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"Flushed replays successfully"];
            [self removeReplaysFromBuffer:replaysArray];
            [Radar flushLogs];

        } else {
            if (replayParams) {
                [self writeNewReplayToBuffer:newReplayParams];
            }
        }

        [self setIsFlushing:NO];
        if (completionHandler) {
            completionHandler(status, res);
        }
    }];
}

- (void)setIsFlushing:(BOOL)flushing {
    isFlushing = flushing;
}

- (void)clearBuffer {
    [mutableReplayBuffer removeAllObjects];

    // remove persisted replays
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:@"radar-replays"];
}

- (void)removeReplaysFromBuffer:(NSArray<RadarReplay *> *)replays {
    [mutableReplayBuffer removeObjectsInArray:replays];

    // persist the updated buffer
    NSError *error;
    NSData *replaysData = [NSKeyedArchiver archivedDataWithRootObject:mutableReplayBuffer requiringSecureCoding:YES error:&error];
    if (error) {
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"Error archiving replays"]; 
        return;
    } else {
        [[NSUserDefaults standardUserDefaults] setObject:replaysData forKey:@"radar-replays"];
    }
}

- (void)loadReplaysFromPersistentStore {
    NSData *replaysData = [[NSUserDefaults standardUserDefaults] objectForKey:@"radar-replays"];
    if (replaysData) {
        NSError *error;
        NSSet *allowedClasses = [NSSet setWithObjects:[NSArray class], [RadarReplay class], [NSDictionary class], [NSString class], [NSNumber class], nil];
        NSArray<RadarReplay *> *replays = [NSKeyedUnarchiver unarchivedObjectOfClasses:allowedClasses fromData:replaysData error:&error];
        if (error) {
            [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"Error unarchiving replays"]; 
            return;
        }
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:[NSString stringWithFormat:@"Loaded replays | length = %lu", (unsigned long)[replays count]]];
        mutableReplayBuffer = [NSMutableArray arrayWithArray:replays];
    }
}

- (void)dropOldestReplay {
    [mutableReplayBuffer removeObjectsInRange:NSMakeRange(0, 1)];
}

#pragma mark - Batch Methods

- (void)addToBatch:(NSMutableDictionary *)params options:(RadarTrackingOptions *)options {
    BOOL wasEmpty = (mutableReplayBuffer.count == 0);
    
    NSMutableDictionary *batchParams = [params mutableCopy];
    batchParams[@"replayed"] = @(YES);
    long nowMs = (long)([NSDate date].timeIntervalSince1970 * 1000);
    batchParams[@"updatedAtMs"] = @(nowMs);
    [batchParams removeObjectForKey:@"updatedAtMsDiff"];
    
    [self writeNewReplayToBuffer:batchParams];
    
    if (wasEmpty) {
        batchStartTime = [NSDate date];
        
        if (options.batchInterval > 0) {
            [self scheduleBatchTimerWithInterval:options.batchInterval];
        }
    }
    
    [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug
                                       message:[NSString stringWithFormat:@"Added to batch | size = %lu",
                                                (unsigned long)mutableReplayBuffer.count]];
}

- (BOOL)shouldFlushBatchWithOptions:(RadarTrackingOptions *)options {
    if (mutableReplayBuffer.count == 0) {
        return NO;
    }
    
    if (options.batchSize > 0 && mutableReplayBuffer.count >= options.batchSize) {
        [[RadarLogger sharedInstance] logWithLevel: RadarLogLevelDebug
                                           message: @"Batch size limit reached"];
        
        return YES;
    }
    
    return NO;
}

- (void)scheduleBatchTimerWithInterval:(int)interval {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self->batchFlushTimer) {
            [self->batchFlushTimer invalidate];
            self->batchFlushTimer = nil;
        }
        
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug
                                           message:[NSString stringWithFormat:@"Scheduling batch timer | interval = %d", interval]];
        
        self->batchFlushTimer = [NSTimer scheduledTimerWithTimeInterval:interval
                                                                repeats:NO
                                                                  block:^(NSTimer *_Nonnull timer) {
            [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug
                                               message:@"Batch timer fired"];
            self->batchStartTime = nil;
            [self flushReplaysWithCompletionHandler:nil completionHandler:nil];
        }];
    });
}

- (void)cancelBatchTimer {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self->batchFlushTimer) {
            [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug
                                               message:@"Canceling batch timer"];
            [self->batchFlushTimer invalidate];
            self->batchFlushTimer = nil;
        }
    });
}

- (void)flushBatch {
    [self cancelBatchTimer];
    batchStartTime = nil;
    [self flushReplaysWithCompletionHandler:nil completionHandler:nil];
}

- (NSUInteger)batchCount {
    return mutableReplayBuffer.count;
}

@end

File: RadarSDK/RadarReplayBuffer.swift
//
//  RadarReplayBuffer.swift
//  RadarSDK
//
//  Created by Alan Charles on 6/29/26.
//  Copyright © 2026 Radar Labs, Inc. All rights reserved.
//

import Foundation

@objc(RadarReplayBuffer)
final class RadarReplayBuffer: NSObject, @unchecked Sendable {
    
    private static let maxBufferSize = 120 // one hour of updates
    private static let storageKey = "radar-replays"
    
    var mutableReplayBuffer: [RadarReplay] = []
    private var isFlushing = false
    private var batchFlushTimer: Timer?
    
    @objc(sharedInstance)
    static let sharedInstance = RadarReplayBuffer()
    
    private override init() {
        super.init()
    }
    
    @objc
    var flushableReplays: [RadarReplay] {
        return mutableReplayBuffer
    }
    
    @objc(writeNewReplayToBuffer:)
    func writeNewReplayToBuffer(_ replayParams: [AnyHashable: Any]) {
        if mutableReplayBuffer.count >= Self.maxBufferSize {
            dropOldestReplay()
        }
        
        let radarReplay = RadarReplay(params: replayParams)
        mutableReplayBuffer.append(radarReplay)
        
        guard let sdkConfiguration = RadarSettings.sdkConfiguration, sdkConfiguration.usePersistence else {
            return
        }
        
        // if buffer length is above 50, remove every fifth replay from the persisted buffer
        let bufferToPersist: [RadarReplay]
        if mutableReplayBuffer.count > 50 {
            bufferToPersist = mutableReplayBuffer.enumerated()
                .filter { ($0.offset + 1) % 5 != 0 }
                .map { $0.element }
        } else {
            bufferToPersist = mutableReplayBuffer
        }
        
        do {
            let replaysData = try NSKeyedArchiver.archivedData(withRootObject: bufferToPersist, requiringSecureCoding: true)
            UserDefaults.standard.set(replaysData, forKey: Self.storageKey)
        } catch {
            RadarLogger.shared.debug("Error archiving replays")
        }
    }
    
    @objc(flushReplaysWithCompletionHandler:completionHandler:)
    func flushReplays(withCompletionHandler replayParams: [AnyHashable: Any]?, completionHandler: RadarFlushReplaysCompletionHandler?) {
        if isFlushing {
            RadarLogger.shared.debug("Already flushing replays")
            completionHandler?(.errorServer, nil)
            return
        }
        
        isFlushing = true
        
        let replaysArray = flushableReplays
        if replaysArray.isEmpty && replayParams == nil {
            RadarLogger.shared.debug("No replays to flush")
            isFlushing = false
            completionHandler?(.success, nil)
            return
        }
        
        var replaysRequestArray = replaysArray.map { $0.replayParams }
        
        // if we have a current track update, add it to the local replay list
        var newReplayParams: [AnyHashable: Any]?
        if let replayParams = replayParams {
            var params = replayParams
            params["replayed"] = true
            params["updatedAtMs"] = Int(Date().timeIntervalSince1970 * 1000)
            // rely on updatedAtMs for replays, not updatedAtMsDiff
            params.removeValue(forKey: "updatedAtMsDiff")
            newReplayParams = params
            replaysRequestArray.append(params)
        }
        
        RadarLogger.shared.debug("Flushing \(replaysRequestArray.count) replays")
        
        guard let bridge = RadarSwift.bridge else {
            isFlushing = false
            completionHandler?(.errorServer, nil)
            return
        }
        
        bridge.flushReplaysRequest(replaysRequestArray) { [self] status, res in
            if status == .success {
                RadarLogger.shared.debug("Flushed replays successfully")
                removeReplaysFromBuffer(replaysArray)
                RadarLogger.flushLogs()
            } else if replayParams != nil, let newReplayParams = newReplayParams {
                writeNewReplayToBuffer(newReplayParams)
            }
            
            setIsFlushing(false)
            completionHandler?(status, res)
        }
    }
    
    @objc(setIsFlushing:)
    func setIsFlushing(_ flushing: Bool) {
        isFlushing = flushing
    }
    
    @objc
    func clearBuffer() {
        mutableReplayBuffer.removeAll()
        UserDefaults.standard.removeObject(forKey: Self.storageKey)
    }
    
    private func removeReplaysFromBuffer(_ replays: [RadarReplay]) {
        mutableReplayBuffer.removeAll { replay in replays.contains { $0.isEqual(replay) } }
        
        do {
            let replaysData = try NSKeyedArchiver.archivedData(withRootObject: mutableReplayBuffer, requiringSecureCoding: true)
            UserDefaults.standard.set(replaysData, forKey: Self.storageKey)
        } catch {
            RadarLogger.shared.debug("Error archiving replays")
        }
    }
    
    @objc
    func loadReplaysFromPersistentStore() {
        guard let replaysData = UserDefaults.standard.object(forKey: Self.storageKey) as? Data else {
            return
        }
        
        let allowedClasses: [AnyClass] = [NSArray.self, RadarReplay.self, NSDictionary.self, NSString.self, NSNumber.self]
        do {
            let replays = try NSKeyedUnarchiver.unarchivedObject(ofClasses: allowedClasses, from: replaysData) as? [RadarReplay]
            RadarLogger.shared.debug("Loaded replays | length = \(replays?.count ?? 0)")
            if let replays = replays {
                mutableReplayBuffer = replays
            }
        } catch {
            RadarLogger.shared.debug("Error unarchiving replays")
        }
    }
    
    @objc
    func dropOldestReplay() {
        if !mutableReplayBuffer.isEmpty {
            mutableReplayBuffer.removeFirst()
        }
    }
    
    // MARK: - Batch Methods
    
    @objc(addToBatch:options:)
    func addToBatch(_ params: [AnyHashable: Any], options: RadarTrackingOptions) {
        var batchParams = params
        batchParams["replayed"] = true
        batchParams["updatedAtMs"] = Int(Date().timeIntervalSince1970 * 1000)
        batchParams.removeValue(forKey: "updatedAtMsDiff")
        
        writeNewReplayToBuffer(batchParams)
        
        if options.batchInterval > 0 && batchFlushTimer == nil {
            scheduleBatchTimer(withInterval: options.batchInterval)
        }
        
        RadarLogger.shared.debug("Added to batch | size = \(mutableReplayBuffer.count)")
    }
    
    @objc(shouldFlushBatchWithOptions:)
    func shouldFlushBatch(withOptions options: RadarTrackingOptions) -> Bool {
        if mutableReplayBuffer.isEmpty {
            return false
        }
        
        if options.batchSize > 0 && mutableReplayBuffer.count >= Int(options.batchSize) {
            RadarLogger.shared.debug("Batch size limit reached")
            return true
        }
        
        return false
    }
    
    @objc(scheduleBatchTimerWithInterval:)
    func scheduleBatchTimer(withInterval interval: Int32) {
        DispatchQueue.main.async {
            if let timer = self.batchFlushTimer {
                timer.invalidate()
                self.batchFlushTimer = nil
            }
            
            RadarLogger.shared.debug("Scheduling batch timer | interval = \(interval)")
            
            self.batchFlushTimer = Timer.scheduledTimer(withTimeInterval: TimeInterval(interval), repeats: false) { _ in
                RadarLogger.shared.debug("Batch timer fired")
                self.batchFlushTimer = nil
                self.flushBatch()
            }
        }
    }
    
    @objc
    func cancelBatchTimer() {
        let cancelTimerBlock = {
            if let timer = self.batchFlushTimer {
                RadarLogger.shared.debug("Canceling batch timer")
                timer.invalidate()
                self.batchFlushTimer = nil
            }
        }
        
        if Thread.isMainThread {
            cancelTimerBlock()
        } else {
            DispatchQueue.main.sync(execute: cancelTimerBlock)
        }
    }
    
    @objc
    func flushBatch() {
        if isFlushing {
            RadarLogger.shared.debug("Skipping batch flush; already flushing replays")
            return
        }
        cancelBatchTimer()
        flushReplays(withCompletionHandler: nil, completionHandler: nil)
    }
    
    @objc
    func batchCount() -> UInt {
        return UInt(mutableReplayBuffer.count)
    }
}
